### DISCUSSION TITLE
How to localize `Info.plist` entries?

### DISCUSSION THREAD

**Discussion by ConfusedVorlon** [FlineDev/RemafoX#61] [Help / FAQ] [answered]
Is there a way to use RemafoX for `Info.plist` translations?
e.g. `NSBluetoothAlwaysUsageDescription`

**Comment by Jeehut** [comment, marked as answer]
@ConfusedVorlon Thank you for this question.

The short answer is:
**Yes**, you can totally use RemafoX to localize `Info.plist` files!
But you need to first localize your `Info.plist` file in general, much like with any Storyboard/XIB file, too.

Apple's documentation on localizing `Info.plist` files is quite hidden so I think this is a very common question. But you can actually find it, in the ["About Information Property List Files"](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/AboutInformationPropertyListFiles.html) page under the section titled "Localizing Property List Values". The most important bit of info is this:

> Localized values are not stored in the `Info.plist` file itself. Instead, you store the values for a particular localization in a strings file with the name `InfoPlist.strings`.

### Step-by-step Guide

So, all you need to do is to:

1. Create a new Strings file in your project and name it exactly `InfoPlist.strings`
2. Hit the "Localize" button on the right sidebar of Xcode
3. Add a file for each language you want to support in the same right sidebar
4. Add all keys and their values for your source language of the `InfoPlist.strings` file

Here are a few screenshots showing the creation steps:
<img width="742" alt="Screenshot 2022-11-15 at 22 51 07" src="https://user-images.githubusercontent.com/6942160/201938074-30b714ff-f3e8-435c-b572-de0c6e6ca91f.png">
<img width="840" alt="Screenshot 2022-11-15 at 22 51 22" src="https://user-images.githubusercontent.com/6942160/201938063-1812a1b2-a9b7-4089-85ff-d1f848efc971.png">
<img width="383" alt="Screenshot 2022-11-15 at 22 51 56" src="https://user-images.githubusercontent.com/6942160/201938076-fd74af53-8e76-4545-a774-1a87d505f201.png">

The result should look something like this:
<img width="1588" alt="Screenshot 2022-11-15 at 22 49 20" src="https://user-images.githubusercontent.com/6942160/201938163-81ce3f0f-279c-47b4-b960-d7204607b350.png">

### Using RemafoX for the Info.plist file

Now, you have a set of `.strings` files for your `Info.plist` entries. Assuming that your search paths in RemafoX match the folder you created the `InfoPlist.strings` files in (if not, adjust them accordingly), you can now use all RemafoX features, including:

1. Opening the "Normalizer" pane in your config and hitting "Normalize Strings files" to sync your keys to all languages
2. Opening the "Quick Actions" or "Machine Translation" pane and hitting "Start batch-translation" to translate to all languages
3. Or building your project to execute the above two actions (Normalize + Machine-Translate) via the build-script automatically

For example, this was the result when I added the `NSBluetoothAlwaysUsageDescription` to a sample project and hit "Start batch-translation" in the config file:
<img width="1072" alt="Screenshot 2022-11-15 at 22 49 58" src="https://user-images.githubusercontent.com/6942160/201938425-8d91e9ed-2fc8-4013-8834-aefe961f245f.png">

### Further Notes

I'm not really happy with how Xcode is dealing with the localization of `Info.plist`. It's easy when you know it, but your question is proof that it's not very well documented, and it's definitely not intuitive either. So I actually have [this feature request](https://github.com/FlineDev/RemafoX/issues/35) to make it even easier in a future update. Feel free to vote it up so I can prioritize it accordingly.

Please let me know if you have any further questions.

**Reply by ConfusedVorlon** [reply]
thank you :)

**Reply by Jeehut** [reply]
You're welcome!
